### PR TITLE
[infra] election

### PR DESCRIPTION
**Pull request by 0x6770**
A draft implementation of leader election by single-vote plurality

### CODE REVIEW

**Comment by neeldug** [comment]
Quick question, are we enforcing leader election through the infrastructure? I figured we landed on allowing the social constructs to exist solely on the peer to peer communications, but perhaps we want to change directions?

**Comment by neeldug** [comment]
Moreover are we going to provide unilateral decision-making that enforces actions to the infrastructure’s provided leadership mechanism.

**Comment by 0x6770** [comment]
> Quick question, are we enforcing leader election through the infrastructure? I figured we landed on allowing the social constructs to exist solely on the peer to peer communications, but perhaps we want to change directions?

These problems are still open to discussion I guess? But I think implement leader election or some part of the governance system in infrastructure is the easier and more robust way.

**Comment by neeldug** [comment]
> > Quick question, are we enforcing leader election through the infrastructure? I figured we landed on allowing the social constructs to exist solely on the peer to peer communications, but perhaps we want to change directions?
> 
> These problems are still open to discussion I guess? But I think implement leader election or some part of the governance system in infrastructure is the easier and more robust way.

I think I agree, however, by decoupling the organisation from the game implementation, it provides an easier maintenance with regards to the core infrastructure and potentially more interesting scenarios where rather than leaders being elected, governance can form independently, essentially treating the game as a black box and the agents the opportunity to organise however they wish.

**Comment by 0x6770** [comment]
I remember Jeremy mentioned in last week's lecture that we could model each level of the game into several rounds, e.g. governance, combat, allocation and trading. 

We could potentially implement all of these base on p2p, but it feels a bit over complicated. 

The p2p system will only work if the agents agree on some common rules which could be simply provided by infra.

**Comment by neeldug** [comment]
Afaik, this could be the case but implementing it on the agent side provides scope for more creative and flexible strategies regarding ignoring mandate and no centralisation of power, which could be mapped out, and given that the election communications ought not to interfere with the game loop, I figured it could just be a messaging type rather than a separate round within the game, I.e. the option to elect a leader should be optional and not bound by any voting system until some sort of convergence occurs.

**Comment by 0x6770** [comment]
> > > Quick question, are we enforcing leader election through the infrastructure? I figured we landed on allowing the social constructs to exist solely on the peer to peer communications, but perhaps we want to change directions?
> 
> > 
> 
> > These problems are still open to discussion I guess? But I think implement leader election or some part of the governance system in infrastructure is the easier and more robust way.
> 
> 
> 
> I think I agree, however, by decoupling the organisation from the game implementation, it provides an easier maintenance with regards to the core infrastructure and potentially more interesting scenarios where rather than leaders being elected, governance can form independently, essentially treating the game as a black box and the agents the opportunity to organise however they wish. 
> 
> 
> 
> Also, would by providing an ability to govern from the infrastructure imply that all agents must comply to this specific elected leadership regime?

Yep I agree, organise election by infra put a lot restriction on how agent could behave. My concern about decentralised election and allocation is that, we may not be able to prevent agent from disobey rules set by the majority of elected authorities. It could develop into a case which all agents race to be the first to send msg back to infra so that it can rob everything during allocation. Or we need to find a way to punish these bad agents?

**Comment by neeldug** [comment]
Yep — so I think we discussed dealing with bad actors who disobey the mandate or provide misrepresentations of their proposed actions, in by that all actions are shown in the follow-up round to enable all agents to determine bad actors. These can then be combatted by formed governments as they can have expectations of higher variance on these bad actors, and as the infrastructure we could try implementing a fair way to allow agents to impose penalties perhaps? But in reality, just by allowing agents to account for bad actors, eventually they should end up being pruned or have a lower average metric score.

**Comment by 0x6770** [comment]
> Yep — so I think we discussed dealing with bad actors who disobey the mandate or provide misrepresentations of their proposed actions, in by that all actions are shown in the follow-up round to enable all agents to determine bad actors. These can then be combatted by formed governments as they can have expectations of higher variance on these bad actors, and as the infrastructure we could try implementing a fair way to allow agents to impose penalties perhaps? But in reality, just by allowing agents to account for bad actors, eventually they should end up being pruned or have a lower average metric score.
> 
> Lastly, with regards to the speed race, this shouldn’t exist since we’ll randomly select when it comes to common pool resource allocation that conflicts, which should enable fairness and mutiny opportunities

I see. I realized I miss a few things from last infra meeting. Shall we start documenting these ideas/rules even just draft? Agent teams will need them to work out their strategy. I can help to doc them e.g.

- How to form a government/organize an election if needed
- Are resources allocated randomly, based on some rules, or by elected leaders

**Comment by neeldug** [comment]
> I see. I realized I miss a few things from last infra meeting. Shall we start documenting these ideas/rules even just draft? Agent teams will need them to work out their strategy. I can help to doc them e.g.
> 
> * How to form a government/organize an election if needed
> * Are resources allocated randomly, based on some rules, or by elected leaders

Yep, that would be great since I think formalising it into a document would ensure everyone's up-to-date with regards to the direction of the gameplay. I think specifically, the governance/organisation will be a social construct that is unenforced by the game itself, so these messages and electorate system needs to be decided between the agent teams and we should standardise the message formats in order to aid this. In terms of the resource allocation, in some earlier discussions we had, we found it could be quite interesting if greedy agents could exist, i.e. stealing the maximum value items from the pool - however, essentially each agent could make a loot decision when it comes to the common pool resource after some p2p communication to choose how to distribute the loot, and in the case of a tie, i.e. a conflicting loot choice between multiple agents, this would be broken randomly and it would be known to all agents who ended up receiving what loot in order to identify who's a bad actor.

**Comment by 0x6770** [comment]
Do we have a plan on why a leader might be needed? maybe we should work out resource allocation first, I just created #38 for further discussion on that topic.

**Comment by Razvan00Rusu** [comment]
Continuing work on this topic based on conversations in #59

File: pkg/infra/game/decision/decision.go [hunk 1]
```diff
@@ -19,3 +19,18 @@ const (
...
+	FightImposition bool
+	LootImposition  bool
+	ResignThreshold uint
+	TermLength      uint
+}
+
+type VotingStrategy uint
+
+const (
+	ConfidenceVote uint = iota
```
Comment (Razvan00Rusu) [inline pkg/infra/game/decision/decision.go:34]: This needs to be taken out of here as a confidence vote isn't an election strategy but will do that tmrw.

**Review by neeldug — APPROVED** [review]
LGTM & fixes #59 in part

**Review by Razvan00Rusu — APPROVED** [review]

**Pull request merged**